dotnet/samples/KernelSyntaxExamples/Example59_OpenAIFunctionCalling.cs

```diff
@@ -0,0 +1,111 @@
+using Microsoft.SemanticKernel;
+using Microsoft.SemanticKernel.AI.ChatCompletion;
+using Microsoft.SemanticKernel.Connectors.AI.OpenAI;
+using Microsoft.SemanticKernel.Connectors.AI.OpenAI.AzureSdk;
+using Microsoft.SemanticKernel.Functions.OpenAPI.Extensions;
+using Microsoft.SemanticKernel.Orchestration;
+using Microsoft.SemanticKernel.Plugins.Core;
+using RepoUtils;
+
+/**
+ * This example shows how to use OpenAI's function calling capability via the chat completions interface.
+ * For more information, see https://platform.openai.com/docs/guides/gpt/function-calling.
+ */
+
+// ReSharper disable once InconsistentNaming
+public static class Example59_OpenAIFunctionCalling
+{
+    public static async Task RunAsync()
+    {
+        IKernel kernel = await InitializeKernelAsync();
+        var chatCompletion = kernel.GetService<IChatCompletion>();
+        var chatHistory = chatCompletion.CreateNewChat();
+
+        OpenAIRequestSettings requestSettings = new()
+        {
+            // Include all functions registered with the kernel.
+            // Alternatively, you can provide your own list of OpenAIFunctions to include.
+            Functions = kernel.Functions.GetFunctionViews().Select(f => f.ToOpenAIFunction()).ToList(),
+        };
+
+        // Set FunctionCall to the name of a specific function to force the model to use that function.
+        requestSettings.FunctionCall = "TimePlugin-Date";
+        await CompleteChatWithFunctionsAsync("What day is today?", chatHistory, chatCompletion, kernel, requestSettings);
+
+        // Set FunctionCall to auto to let the model choose the best function to use.
+        requestSettings.FunctionCall = OpenAIRequestSettings.FunctionCallAuto;
+        await CompleteChatWithFunctionsAsync("What computer tablets are available for under $200?", chatHistory, chatCompletion, kernel, requestSettings);
+    }
+
+    private static async Task<IKernel> InitializeKernelAsync()
+    {
+        // Create kernel with chat completions service
+        IKernel kernel = new KernelBuilder()
+            .WithLoggerFactory(ConsoleLogger.LoggerFactory)
+            .WithOpenAIChatCompletionService(TestConfiguration.OpenAI.ChatModelId, TestConfiguration.OpenAI.ApiKey, serviceId: "chat")
+            //.WithAzureChatCompletionService(TestConfiguration.AzureOpenAI.ChatDeploymentName, TestConfiguration.AzureOpenAI.Endpoint, TestConfiguration.AzureOpenAI.ApiKey, serviceId: "chat")
+            .Build();
+
+        // Load functions to kernel
+        kernel.ImportFunctions(new TimePlugin(), "TimePlugin");
+        await kernel.ImportPluginFunctionsAsync("KlarnaShoppingPlugin", new Uri("https://www.klarna.com/.well-known/ai-plugin.json"), new OpenApiFunctionExecutionParameters());
+
+        return kernel;
+    }
+
+    private static async Task CompleteChatWithFunctionsAsync(string ask, ChatHistory chatHistory, IChatCompletion chatCompletion, IKernel kernel, OpenAIRequestSettings requestSettings)
```

Comment:
What's the plan for including a variant of this "in the box"?  It's boilerplate that lots of folks are going to need to write, otherwise.

Reply:
Tracked by #2932. Here is one potential way we could do this, would love your feedback there @stephentoub https://github.com/gitri-ms/semantic-kernel/pull/1

Reply:
There's some issues with the SKChatMessage and MessageBase models that's actually not allowing the full scenario here. Need to figure out how to resolve that.

Reply:
Sorry, I missed these responses somehow.

Yes, something along the lines of https://github.com/gitri-ms/semantic-kernel/pull/1.  I'm not seeing a need for it to return the ChatHistory, since it looks like it's always just returning the one that was passed in.

From my perspective, this support is critical and is _the_ value add SK can have here.  It also just "makes sense": SK already has the functions, and has the metadata for them, and that metadata is being used to populate the functions description, and it knows how to invoke its functions, and it knows how to convert the arguments supplied by the LLM to the right types for the function invocation, and it knows how to marshal the results back... it has everything it needs to make what is otherwise complicated and cumbersome boilerplate into a one-liner.